File: agent-packs/eval-pilot/engine-ts/src/spec.ts
/**
 * The {@link EvalSpec} — the compiled, engine-neutral description of one eval.
 *
 * Both authoring surfaces (the Markdown `*.eval.md` loader and the fluent
 * TypeScript builder) produce an {@link EvalSpec}. The executor consumes it.
 * Nothing here executes anything; this module is pure data plus a little
 * validation.
 *
 * Structure mirrors arrange → act → assert.
 */

export const AGENT = "agent";
export const SKILL = "skill";
export const NONE = "none";

export type EvalKind = typeof AGENT | typeof SKILL | typeof NONE;

/** What agents/skills to copy into the eval workspace. */
export interface StageSpec {
  agent: string | null;
  skill: string | null;
  all: boolean;
  include_skills: boolean;
}

export function makeStageSpec(init: Partial<StageSpec> = {}): StageSpec {
  return {
    agent: null,
    skill: null,
    all: false,
    include_skills: true,
    ...init,
  };
}

/** Copy fixture files into the workspace before the SUT runs. */
export interface FileCopy {
  copy: string; // source path/glob, relative to the spec's base_dir
  dest: string; // destination subdir inside the workspace
}

export function makeFileCopy(copy: string, dest = "."): FileCopy {
  return { copy, dest };
}

/** Arrange step: staging + fixture files. */
export interface SetupSpec {
  stage: StageSpec;
  files: FileCopy[];
}

export function makeSetupSpec(init: Partial<SetupSpec> = {}): SetupSpec {
  return {
    stage: init.stage ?? makeStageSpec(),
    files: init.files ?? [],
  };
}

/** Act step: drive the SUT with a prompt. */
export interface ActSpec {
  prompt: string;
  agent: string | null;
  skill: string | null;
  timeout: number | null;
}

export function makeActSpec(
  init: Partial<ActSpec> & Pick<ActSpec, "prompt">,
): ActSpec {
  return {
    agent: null,
    skill: null,
    timeout: null,
    ...init,
  };
}

/**
 * One assertion. `kind` selects a registered checker; `args` are its
 * parameters. `predicate` carries a callable for `kind='custom'`.
 */
export interface AssertionSpec {
  kind: string;
  args: Record<string, any>;
  name: string | null;
  predicate: ((ctx: any) => any) | null;
}

export function makeAssertionSpec(
  init: Partial<AssertionSpec> & Pick<AssertionSpec, "kind">,
): AssertionSpec {
  return {
    args: {},
    name: null,
    predicate: null,
    ...init,
  };
}

/** An LLM-as-judge assertion over an artifact (or stdout). */
export interface JudgeSpec {
  criteria: string;
  artifact: string | null; // file glob; null => use stdout
  threshold: number;
  name: string;
  golden: string[];
}

export function makeJudgeSpec(
  init: Partial<JudgeSpec> & Pick<JudgeSpec, "criteria">,
): JudgeSpec {
  return {
    artifact: null,
    threshold: 0.7,
    name: "judge",
    golden: [],
    ...init,
  };
}

/**
 * A numeric metric recorded to JSONL history and (optionally) gated.
 *
 * `value` may be a number, a callable (builder only), or a `$`-reference
 * string resolved by the executor: `$judge.score`, `$judge.<name>.score`,
 * `$duration`, `$stdout.words`, `$stdout.chars`, `$assertions.pass_rate`.
 */
export interface MetricSpec {
  name: string;
  value: number | string | ((ctx: any) => number);
  direction: string;
  unit: string;
  baseline_strategy: string;
  baseline: number | null;
  window: number;
  tolerance: number;
  tolerance_pct: number;
  gate: boolean;
}

export function makeMetricSpec(
  init: Partial<MetricSpec> & Pick<MetricSpec, "name" | "value">,
): MetricSpec {
  return {
    direction: "higher_is_better",
    unit: "",
    baseline_strategy: "last",
    baseline: null,
    window: 5,
    tolerance: 0.0,
    tolerance_pct: 0.0,
    gate: false,
    ...init,
  };
}

/** The full compiled description of one eval. */
export interface EvalSpec {
  name: string;
  target: string | null;
  kind: EvalKind;
  tags: string[];
  timeout: number;
  summary: string;
  description: string;
  setup: SetupSpec;
  act: ActSpec | null;
  assertions: AssertionSpec[];
  judges: JudgeSpec[];
  metrics: MetricSpec[];
  spec_path: string | null;
  base_dir: string | null;
}

/**
 * Build an {@link EvalSpec}, applying the same post-init normalization as the
 * Python dataclass `__post_init__` (kind validation + target-staging inference).
 */
export function makeEvalSpec(
  init: Partial<EvalSpec> & Pick<EvalSpec, "name">,
): EvalSpec {
  const spec: EvalSpec = {
    name: init.name,
    target: init.target ?? null,
    kind: init.kind ?? NONE,
    tags: init.tags ?? [],
    timeout: init.timeout ?? 600.0,
    summary: init.summary ?? "",
    description: init.description ?? "",
    setup: init.setup ?? makeSetupSpec(),
    act: init.act ?? null,
    assertions: init.assertions ?? [],
    judges: init.judges ?? [],
    metrics: init.metrics ?? [],
    spec_path: init.spec_path ?? null,
    base_dir: init.base_dir ?? null,
  };

  if (spec.kind !== AGENT && spec.kind !== SKILL && spec.kind !== NONE) {
    throw new Error(
      `eval ${JSON.stringify(spec.name)}: kind must be 'agent', 'skill', or ` +
        `'none', got ${JSON.stringify(spec.kind)}`,
    );
  }

  // Infer the staged target when the setup doesn't say otherwise.
  if (
    spec.target &&
    spec.kind === AGENT &&
    !spec.setup.stage.agent &&
    !spec.setup.stage.all
  ) {
    spec.setup.stage.agent = spec.target;
  }
  if (
    spec.target &&
    spec.kind === SKILL &&
    !spec.setup.stage.skill &&
    !spec.setup.stage.all
  ) {
    spec.setup.stage.skill = spec.target;
  }
  return spec;
}

/** True when the eval carries human-readable intent (summary/description). */
export function specHasProse(spec: EvalSpec): boolean {
  return Boolean(spec.summary.trim() || spec.description.trim());
}

/**
 * True for a structural eval: `kind: none` with no act prompt. These run
 * offline against the repo tree (no SUT) and only need assertions/judges/metrics.
 */
export function specIsStructural(spec: EvalSpec): boolean {
  return spec.kind === NONE && spec.act === null;
}

/** True when the eval has an act prompt and at least one check. */
export function specIsImplemented(spec: EvalSpec): boolean {
  const checksOk = Boolean(
    spec.assertions.length || spec.judges.length || spec.metrics.length,
  );
  if (specIsStructural(spec)) return checksOk;
  const actOk = spec.act !== null && Boolean(spec.act.prompt.trim());
  return actOk && checksOk;
}

/** A description-first draft: has prose but isn't implemented yet. */
export function specIsStub(spec: EvalSpec): boolean {
  return specHasProse(spec) && !specIsImplemented(spec);
}

/** Return a list of human-readable problems (empty when the spec is ok). */
export function validateSpec(spec: EvalSpec): string[] {
  const problems: string[] = [];
  if (!spec.name) problems.push("missing 'name'");
  const structural = specIsStructural(spec);
  if (!structural) {
    if (spec.act === null) {
      problems.push("missing an '## Act' prompt");
    } else if (!spec.act.prompt.trim()) {
      problems.push("act prompt is empty");
    }
  }
  if (spec.kind === AGENT && !(spec.setup.stage.agent || spec.setup.stage.all)) {
    problems.push("kind=agent but no agent is staged (set 'target')");
  }
  if (spec.kind === SKILL && !(spec.setup.stage.skill || spec.setup.stage.all)) {
    problems.push("kind=skill but no skill is staged (set 'target')");
  }
  if (!(spec.assertions.length || spec.judges.length || spec.metrics.length)) {
    problems.push("no assertions, judges, or metrics — nothing to check");
  }
  for (const m of spec.metrics) {
    if (!m.name) problems.push("a metric is missing 'name'");
  }
  return problems;
}

File: agent-packs/eval-pilot/engine/src/evalpilot/spec.py
"""The :class:`EvalSpec` — the compiled, engine-neutral description of one eval.

Both authoring surfaces (the Markdown ``*.eval.md`` loader and the fluent
Python builder) produce an :class:`EvalSpec`. The :mod:`evalpilot.executor`
consumes it. Nothing here executes anything; this module is pure data plus a
little validation, so a spec can be inspected, printed, or lint-checked
without a live SUT.

Structure mirrors arrange → act → assert:

* :class:`SetupSpec`   — what to stage into the workspace (arrange).
* :class:`ActSpec`     — the prompt to drive the SUT with (act).
* :class:`AssertionSpec` / :class:`JudgeSpec` / :class:`MetricSpec` — checks
  and recorded numbers (assert).
"""

from __future__ import annotations

import dataclasses
from pathlib import Path
from typing import Any, Callable, Optional

AGENT = "agent"
SKILL = "skill"
NONE = "none"


@dataclasses.dataclass
class StageSpec:
    """What agents/skills to copy into the eval workspace."""

    agent: Optional[str] = None
    skill: Optional[str] = None
    all: bool = False
    include_skills: bool = True


@dataclasses.dataclass
class FileCopy:
    """Copy fixture files into the workspace before the SUT runs."""

    copy: str            # source path/glob, relative to the spec's base_dir
    dest: str = "."      # destination subdir inside the workspace


@dataclasses.dataclass
class SetupSpec:
    """Arrange step: staging + fixture files."""

    stage: StageSpec = dataclasses.field(default_factory=StageSpec)
    files: list = dataclasses.field(default_factory=list)  # list[FileCopy]


@dataclasses.dataclass
class ActSpec:
    """Act step: drive the SUT with a prompt."""

    prompt: str
    agent: Optional[str] = None   # explicit agent override
    skill: Optional[str] = None   # run in skill-isolation mode
    timeout: Optional[float] = None


@dataclasses.dataclass
class AssertionSpec:
    """One assertion. ``kind`` selects a registered checker; ``args`` are its
    parameters. ``predicate`` carries a Python callable for ``kind='custom'``.
    """

    kind: str
    args: dict = dataclasses.field(default_factory=dict)
    name: Optional[str] = None
    predicate: Optional[Callable] = None


@dataclasses.dataclass
class JudgeSpec:
    """An LLM-as-judge assertion over an artifact (or stdout)."""

    criteria: str
    artifact: Optional[str] = None      # file glob; None => use stdout
    threshold: float = 0.7
    name: str = "judge"
    golden: list = dataclasses.field(default_factory=list)


@dataclasses.dataclass
class MetricSpec:
    """A numeric metric recorded to JSONL history and (optionally) gated.

    ``value`` may be a number, a callable (builder only), or a ``$``-reference
    string resolved by the executor: ``$judge.score``, ``$judge.<name>.score``,
    ``$duration``, ``$stdout.words``, ``$stdout.chars``, ``$assertions.pass_rate``.
    """

    name: str
    value: Any
    direction: str = "higher_is_better"
    unit: str = ""
    baseline_strategy: str = "last"
    baseline: Optional[float] = None
    window: int = 5
    tolerance: float = 0.0
    tolerance_pct: float = 0.0
    gate: bool = False


@dataclasses.dataclass
class EvalSpec:
    """The full compiled description of one eval."""

    name: str
    target: Optional[str] = None
    kind: str = NONE
    tags: list = dataclasses.field(default_factory=list)
    timeout: float = 600.0
    summary: str = ""
    description: str = ""
    setup: SetupSpec = dataclasses.field(default_factory=SetupSpec)
    act: Optional[ActSpec] = None
    assertions: list = dataclasses.field(default_factory=list)  # AssertionSpec
    judges: list = dataclasses.field(default_factory=list)       # JudgeSpec
    metrics: list = dataclasses.field(default_factory=list)      # MetricSpec
    spec_path: Optional[str] = None
    base_dir: Optional[Path] = None

    def __post_init__(self) -> None:
        if self.kind not in (AGENT, SKILL, NONE):
            raise ValueError(
                f"eval {self.name!r}: kind must be 'agent', 'skill', or 'none', "
                f"got {self.kind!r}"
            )
        # Infer the staged target when the setup doesn't say otherwise.
        if self.target and self.kind == AGENT and not self.setup.stage.agent \
                and not self.setup.stage.all:
            self.setup.stage.agent = self.target
        if self.target and self.kind == SKILL and not self.setup.stage.skill \
                and not self.setup.stage.all:
            self.setup.stage.skill = self.target

    def has_prose(self) -> bool:
        """True when the eval carries human-readable intent (summary/description)."""
        return bool(self.summary.strip() or self.description.strip())

    def is_implemented(self) -> bool:
        """True when the eval has an act prompt and at least one check."""
        act_ok = self.act is not None and bool(self.act.prompt.strip())
        checks_ok = bool(self.assertions or self.judges or self.metrics)
        return act_ok and checks_ok

    def is_stub(self) -> bool:
        """A description-first draft: has prose but isn't implemented yet.

        Enables the "write the description, then ask an agent to implement"
        workflow — such files are reported as stubs by ``evalpilot lint``
        rather than as hard errors.
        """
        return self.has_prose() and not self.is_implemented()

    def validate(self) -> list[str]:
        """Return a list of human-readable problems (empty when the spec is ok)."""
        problems: list[str] = []
        if not self.name:
            problems.append("missing 'name'")
        if self.act is None:
            problems.append("missing an '## Act' prompt")
        elif not self.act.prompt.strip():
            problems.append("act prompt is empty")
        if self.kind == AGENT and not (
            self.setup.stage.agent or self.setup.stage.all
        ):
            problems.append("kind=agent but no agent is staged (set 'target')")
        if self.kind == SKILL and not (
            self.setup.stage.skill or self.setup.stage.all
        ):
            problems.append("kind=skill but no skill is staged (set 'target')")
        if not (self.assertions or self.judges or self.metrics):
            problems.append("no assertions, judges, or metrics — nothing to check")
        for m in self.metrics:
            if not m.name:
                problems.append("a metric is missing 'name'")
        return problems


__all__ = [
    "AGENT",
    "SKILL",
    "NONE",
    "StageSpec",
    "FileCopy",
    "SetupSpec",
    "ActSpec",
    "AssertionSpec",
    "JudgeSpec",
    "MetricSpec",
    "EvalSpec",
]
